### ISSUE TITLE
Debian template: disable newly (all) installed services by default

### ISSUE THREAD

**Issue by rootkovska** [QubesOS/qubes-issues#2238]
The role of a Qubes template is to _offer_ software (i.e. a filesystem with installed software), not to _run_ it!

It might break some security assumptions if all the software user intends to install in a template (which the user might want to use only on few of the AppVMs based on that software) was also made _running_ in all the AppVMs based on that template.

While we do not allow for inter-VM networking between AppVM by default, neither to reach AppVM from the outside world (thanks to default DNAT-based routing), there still might be edge cases where enabling all services by default might have fatal consequences. E.g. a service that periodically fetches some data, then parses it (and subsequently gets exploited due to some bugs it might have).

AFAIU, the Debian policy of "enable services for all installed software" is in stark contrast with Fedora's default rule which says the opposite.

**Comment by marmarek** [comment]
Related: #2109 #1928 and #1625 (discussion at the end)

**Comment by Kixunil** [comment]
I've been playing with Debian packaging lately and I think I found a good way to resolve this. `deb-systemd-invoke` should be used by all packages. It's a wrapper that asks `/usr/sbin/policy-rc.d` (despide `.d` it's not a directory, but executable!) if it should execute a service.

I suggest creating a simple script at that location which will check the service against a whitelist of allowed services (Qubes daemons etc).

There's a second part to this though and that is enabling services. While my suggestion above would prevent services from running after install, they would still run after next boot. I don't know if there's a nice way to wrap `deb-systemd-helper` which is used for enabling. I suppose diversion would help. The wrapper should mask all newly installed services (except whitelisted ones) and unmask them during AppVM boot.

Still I'm unsure, if all services should be running in all VMs.

**Comment by marmarek** [comment]
Technically there is much better way to do that: systemd preset file which define what services should be automatically enabled - this is invented specifically to allow setting such defaults.

The question here is rather about policy: generally we [don't change upstream policy decisions](https://www.qubes-os.org/faq/#what-is-qubes-attitude-toward-changing-guest-distros) and Debian has more or less "if service is installed, it should be started automatically". Should we override it?

I think we had more detailed discussion about it somewhere, but can't find it now. @unman @adrelanos any opinions?

**Comment by unman** [comment]
I too seem to remember a discussion, but in my current somewhat addled state, cant remember where.

We do stop services on initial creation of templates, and  I don't think that the proposed solution is of great help given that it takes a hit in the next boot. My gut feeling is that we should stick with Debian default, however painful that is.
If it's decided that we should stop this then we would have to create a wrapper round every install to stop on install and then block by default - not impossible by any means.

**Comment by adrelanos** [comment]
We can't or shouldn't harden Debian (much) since the result wouldn't be a Debian template with only necessarily modifications to integrate with Qubes. That would be a fork of Debian. Debian is a trademark.

https://www.debian.org/trademark

Forks of Debian shouldn't be called Debian to avoid confusion.

Instead, I would suggest to maintain a reasonably hardened by default Debian derivative. Any coming to mind? Perhaps `Kicksecure`?

https://www.whonix.org/wiki/Kicksecure

Anyone up to maintain a Kicksecure Qubes template?

And then once such as template stabilized in Qubes consider installing only that by default and not "plain" Debian. That would avoid the discussion on whether (default less secure) distribution defaults should be honored. No change of upstream policy but Qubes becoming "upstream" by not calling it Debian but its own name that it deserves.

**Comment by Kixunil** [comment]
@marmarek fair, would it be fine to at least disable it in template VM? Possibly provide a very easy way to turn it on (or make it opt-out). *Not* running things like [mpd](https://www.musicpd.org/) in template VM would be nice.

**Comment by adrelanos** [comment]
Potential other alternatives:

* https://packages.debian.org/bookworm/policy-rcd-declarative
* https://packages.debian.org/bookworm/policy-rcd-declarative-allow-all
* https://packages.debian.org/bookworm/policy-rcd-declarative-deny-all

**Comment by ArrayBolt3** [comment]
I don't believe policy-rc.d is actually all that helpful here. It can be used to prevent services from starting or restarting *when packages are installed*, but it doesn't keep services from being started on system bootup (which almost all services will be by default in Debian).

The way Kicksecure is working around this for our sysmaint session is that we implement a new systemd target that is booted into instead of `graphical.target` or `multi-user.target`, and that target only depends on the bare minimum services it needs to function properly. This will be available in general once user-sysmaint-split is ready for use.

Edit: The mechanism described above is now part of Kicksecure 18 and is fully functional on Qubes R4.3 and non-Qubes platforms.

**Comment by ben-grande** [comment]
> I don't believe policy-rc.d is actually all that helpful here. It can be used to prevent services from starting or restarting _when packages are installed_, but it doesn't keep services from being started on system bootup (which almost all services will be by default in Debian).
> 
> The way Kicksecure is working around this for our sysmaint session is that we implement a new systemd target that is booted into instead of `graphical.target` or `multi-user.target`, and that target only depends on the bare minimum services it needs to function properly. This will be available in general once user-sysmaint-split is ready for use.

Have you looked into systemd.preset.5, mentioned by Marek above?

**Comment by ArrayBolt3** [comment]
@ben-grande Yes, and in Kicksecure we use systemd presets for some units that we want to ship but not enable by default. It works, though it raises the question of if it's really a good idea to modify Debian's default configuration. Patrick already mentioned the risk of trademark problems, but I'm concerned about compatibility issues. What if (for instance) a Debian package starts a systemd unit immediately upon installing it, but doesn't bother enabling it because it thinks that Debian already enables it by default? Then the service works on installation, and breaks for no apparent reason after a reboot. With Kicksecure's sysmaint session, this isn't as much of a problem since the user expectation is already that unnecessary services won't run, but for general use it initially sounds like a bad idea to me.

**Comment by ben-grande** [comment]
> The question here is rather about policy: generally we [don't change upstream policy decisions](https://www.qubes-os.org/faq/#what-is-qubes-attitude-toward-changing-guest-distros) and Debian has more or less "if service is installed, it should be started automatically". Should we override it?

Now quoting the docs:

> What is Qubes’ attitude toward changing guest distros?[](https://doc.qubes-os.org/en/latest/introduction/faq.html#what-is-qubes-attitude-toward-changing-guest-distros)
> 
> In general, we try to respect each distro’s culture, but we reserve the right to make modifications that we deem appropriate. See the discussion on issue [#1014](https://github.com/QubesOS/qubes-issues/issues/1014) for an example.
> 
> The policy is there mostly to ease maintenance, on several levels:
> 
>  -   Less modifications means easier migration to new upstream distribution releases.

Needs to be researched on the impact. We will probably run `preset-all` on such dist-upgrades and the presets needs that should be enabled must be clear. At least this is what happens on dom0, so we probably should write the preset and document that system administrators should do it too.

>  -   The upstream documentation matches the distribution running in the Qubes VM.

Wouldn't be true anymore, `systemctl enable UNIT` is not widely documented.

>  -   We’re less likely to introduce Qubes-specific issues.

In this case, this is the same point as the first one.
 
>  -   Each officially supported distribution (ideally) should offer the same set of Qubes-specific features - a change in one supported distribution should be followed also in others, including new future distributions.

---

So no, I don't think we should change Debian's systemd preset policy. It breaks guides and will increase user support.

I think that if someone wants such default disabled preset, they should set it themselves or use a qube based on Fedora template for that. On Debian, I mask each unwanted service before installing the package containing it, at least on Qusal.

But if you plan to change such policy, then I don't think it should be called Debian anymore. It is not simply configuring the system to be well integrated in Qubes, it is hardening the guest.
